gax/src/test/java/com/google/api/gax/batching/v2/FlowControllerTest.java

```diff
@@ -0,0 +1,253 @@
+    } catch (IllegalArgumentException expected) {
+      actualException = expected;
+    }
+    assertThat(actualException).isInstanceOf(IllegalArgumentException.class);
+    flowController.release(1, 0);
+  }
+
+  @Test
+  public void testReserveRelease_noLimits_ok() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    flowController.reserve(1);
+    flowController.release(1, 1);
+  }
+
+  @Test
+  public void testReserveRelease_ignore_ok() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(1)
+                .setMaxOutstandingRequestBytes(1L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    flowController.reserve(1);
+    flowController.release(1, 1);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByElementCount_noBytesLimit() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(2)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+    flowController.reserve(1);
+
+    testBlockingReserveRelease(flowController, 5);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByNumberOfBytes() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(100)
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    testBlockingReserveRelease(flowController, 10);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByNumberOfBytes_noElementCountLimit() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    testBlockingReserveRelease(flowController, 10);
+  }
+
+  private static void testBlockingReserveRelease(
+      final FlowController flowController, final int maxNumBytes) throws Exception {
+
+    flowController.reserve(1);
+
+    final SettableFuture<?> permitsReserved = SettableFuture.create();
+    Future<?> finished =
+        Executors.newCachedThreadPool()
+            .submit(
+                new Runnable() {
+                  @Override
+                  public void run() {
+                    try {
+                      permitsReserved.set(null);
+                      flowController.reserve(maxNumBytes);
```

Comment:
this seems wrong: shouldnt bytes be reserved before the checkpoint future? Otherwise the main thread can call release before this thread calls reserve

Reply:
This was intentional. We have already reserved one  byte at the start so `flowController.reserve(maxNumBytes)` will block until the release of at least one byte.

please let me know if we should update our approach for testing these.